syntax = "proto2";
package netData;

// 2-1 握手
message HandShake
{
	optional bytes sClientKey = 1;		//base64encode(randomkey)
}

// 2-2 握手返回
message HandShakeRet
{
	enum EC
	{
		EC_OTHER = 0;
		EC_SUCC = 1;
		EC_FAIL = 2;
	};
	optional int32 nCode = 1;	//握手结果
	optional bytes sChallenge = 2;		//base64encode(randomkey)
	optional bytes sServerKey = 3;		//base64encode(dhexchange(randomkey))
}

// 收到1-2之后
// 使用 nServerKey 和 nClientKey 计算出 secret
// 使用 secret 加密 nChallenge
// secret = dhsecret(nClientKey,nServerKey)

// 2-3 挑战握手
message Challenge
{
	optional string sHmac = 1;		//base64encode(hmac64(nChallenge,secret))
}

// 2-4 挑战握手返回
message ChallengeRet
{
	enum EC
	{
		EC_OTHER = 0;
		EC_SUCC = 1;
		EC_FAIL = 2;
	};
	optional int32 nCode = 1;	//握手结果
}

// 2-5 认证(账号密码)
message Auth
{
	//token:base64encode(账号)@base64encode(要登陆的服务器ID)@base64encode(密码)
	optional string Account = 1;	//base64encode(desencode(secret,token))
	optional bytes Secret = 2; 	//client不用填写
}

// 2-6 认证返回
message AuthRet
{
	enum EC
	{
		EC_OTHER = 0;
		EC_SUCC = 1;
		EC_FAIL = 2;
		EC_HANDSHAKE = 3;	// 握手失败
		EC_ADDACCOUNT = 4;	// Login添加账号失败
		EC_DB = 5;			// 数据库查询失败
		EC_AUTHINFO = 6;	// 没有找到认证信息
		EC_AUTHING = 7;		// 正在认证中，稍后再试
		EC_KICK = 8;		// 已经将在线玩家踢下线，请重试
		EC_WATING = 9;		// 系统正在踢在线玩家，请稍微
		EC_LOGINTIME = 10;	// 更新账号登陆时间失败
		EC_CREATE = 11;		// 创建账号失败
		EC_DBSTAUTS = 12;	// 数据库没有启动
	};
	optional int32 nCode = 1;
	optional string Account = 2;	//账号
}

message PlayerLite
{
	optional int64 nGuid = 1;
	optional string sName = 2;
	optional int32 nJob = 3;
	optional int32 nSex = 4;
}

// 2-7 请求角色列表
message PlayerList
{
	// 空消息
	optional string Account = 1;	//请求不用填写
}

// 2-8 返回角色列表
message PlayerListRet
{
	repeated PlayerLite list = 1;
}

// 2-9 请求创建角色
message CreatePlayer
{
	enum EC
	{
		EC_OTHER = 0;
		EC_SUCC = 1;
		EC_FAIL = 2;
		EC_REPEATED = 4;//名称重复
	};
	optional string sName = 1;
	optional int32 nJob = 2;
	optional int32 nSex = 3;
	optional string Account = 4;//请求不用填写
	optional int32 nNameCheckRet = 5;	//名称检查服务器返回的结果，请求不用填写
}

// 2-10 创建角色返回
message CreatePlayerRet
{
	enum EC
	{
		EC_OTHER = 0;
		EC_SUCC = 1;
		EC_FAIL = 2;
		EC_CREATE = 3;
		EC_REPEATED = 4;//名称重复
		EC_NAMESVR = 5;//名称检查服务器未连接
	};
	optional int32 nCode = 1;
	optional PlayerLite Info = 2;
}

// 2-11 请求选择角色
message SelectPlayer
{
	optional int64 nGuid = 1;
}

// 2-12 选择角色返回
message SelectPlayerRet
{
	enum EC
	{
		EC_OTHER = 0;
		EC_SUCC = 1;
		EC_FAIL = 2;
		EC_SERVER = 3;	// 服务器未开启
		EC_AUTH = 4;	// 未认证
		EC_NONE = 5;	// 找有找到角色
		EC_LOGINTIME = 6;	// 更新登陆时间失败
	};
	optional int32 nCode = 1;
	optional int32 nMapID = 2;	// 要加载的地图ID
	optional int32 nServerID = 3;// 服务器ID
	optional int32 nLineID = 4;	// 线路ID
	optional int32 nPort = 5;	// 服务器端口
	optional string sIP = 6;		// 需要去连接的服务器ip
}

// 2-13 登陆
message Login
{
	// handshake = base64encode(token.user)@base64encode(token.server)#base64encode(subid):index)
	// hmac = hmac64(hashkey(handshake), secret)
	optional string Account = 1;	// handshake:base64encode(hmac)
	optional bytes Secret = 2;
}

// 2-14 登录返回
message LoginRet
{
	enum EC
	{
		EC_OTHER = 0;
		EC_SUCC = 1;
		EC_FAIL = 2;
		EC_ACCOUNT = 3;	//没有账号信息
		EC_SECRET = 4;	//秘钥错误
		EC_CENTERID = 5;//中心服务器分配id失败
		EC_ADDPLAYER = 6;//逻辑服添加玩家失败
		EC_LOADDATA = 7;//逻辑服加载玩家数据失败
	};
	optional int32 nCode = 1;
	optional int32 nTempID = 2;
}
